File: prototype/rails/src/app/domain/payments/decline_reason.rb
module Domain
  module Payments
    module DeclineReason
      GENERIC_DECLINE = "generic_decline"
      INSUFFICIENT_FUNDS = "insufficient_funds"
      INVALID_CARD_NUMBER = "invalid_card_number"

      ALL = [GENERIC_DECLINE, INSUFFICIENT_FUNDS, INVALID_CARD_NUMBER].freeze

      MESSAGES = {
        GENERIC_DECLINE => "Your card was declined.",
        INSUFFICIENT_FUNDS => "Your card has insufficient funds.",
        INVALID_CARD_NUMBER => "That card number is not valid."
      }.freeze

      module_function

      def message(reason)
        MESSAGES.fetch(reason)
      end
    end
  end
end

File: prototype/php/src/app/Domain/Payments/DeclineReason.php
<?php

namespace App\Domain\Payments;

enum DeclineReason: string
{
    case GenericDecline = 'generic_decline';
    case InsufficientFunds = 'insufficient_funds';
    case InvalidCardNumber = 'invalid_card_number';

    public function message(): string
    {
        return match ($this) {
            self::GenericDecline => 'Your card was declined.',
            self::InsufficientFunds => 'Your card has insufficient funds.',
            self::InvalidCardNumber => 'That card number is not valid.',
        };
    }
}
